pangeo_forge/recipe.py

```diff
@@ -0,0 +1,320 @@
+    along the dimension `sequence_dim`. Each file may contain multiple variables.
+    """
+
+    input_urls: Iterable[str]
+    """The inputs used to generate the dataset."""
+
+    sequence_dim: str
+    """The dimension name along which the inputs will be concatenated."""
+
+    inputs_per_chunk: int = 1
+    """The number of inputs to use in each chunk."""
+
+    nitems_per_input: int = 1
+    """The length of each input along the `sequence_dim` dimension."""
+
+    def __post_init__(self):
+        self._chunks_inputs = {
+            k: v for k, v in enumerate(chunked_iterable(self.input_urls, self.inputs_per_chunk))
+        }
+
+        def drop_vars(ds):
+            # writing a region means that all the variables MUST have sequence_dim
+            to_drop = [v for v in ds.variables if self.sequence_dim not in ds[v].dims]
+            return ds.drop_vars(to_drop)
+
+        self.chunk_preprocess_funcs.append(drop_vars)
+
+    def inputs_for_chunk(self, chunk_key):
+        return self._chunks_inputs[chunk_key]
+
+    def iter_inputs(self):
+        for chunk_key in self.iter_chunks():
+            for input in self.inputs_for_chunk(chunk_key):
+                yield input
+
+    def nitems_for_chunk(self, chunk_key):
+        return self.nitems_per_input * len(self.inputs_for_chunk(chunk_key))
+
+    def region_for_chunk(self, chunk_key):
+        # return a dict suitable to pass to xr.to_zarr(region=...)
+        # specifies where in the overall array to put this chunk's data
+        stride = self.nitems_per_input * self.inputs_per_chunk
+        start = chunk_key * stride
+        return {self.sequence_dim: slice(start, start + self.nitems_for_chunk(chunk_key))}
+
+    def sequence_len(self):
+        # tells the total size of dataset along the sequence dimension
+        return sum([self.nitems_for_chunk(k) for k in self.iter_chunks()])
+
+    def sequence_chunks(self):
+        # chunking
+        return {self.sequence_dim: self.inputs_per_chunk * self.nitems_per_input}
+
+    def iter_chunks(self):
+        for k in self._chunks_inputs:
+            yield k
+
+    @property
+    def prepare(self):
+        def _prepare():
+
+            try:
+                ds = self.open_target()
+                logger.info("Found an existing dataset in target")
+                logger.debug(f"{ds}")
+            except (IOError, zarr.errors.GroupNotFoundError):
+                first_chunk_key = next(self.iter_chunks())
+                ds = self.open_chunk(first_chunk_key).chunk()
+
+                # make sure the concat dim has a valid fill_value to avoid
+                # overruns when writing chunk
+                ds[self.sequence_dim].encoding = {"_FillValue": -1}
+                # actually not necessary if we use decode_times=False
+                self.initialize_target(ds)
+
+            self.expand_target_dim(self.sequence_dim, self.sequence_len())
+
+        return _prepare
+
+
+@dataclass
+class StandardSequentialRecipe(
```

Comment:
Let's think about a more informative name here. Once concern I have is that we'll get to Xarray->Zarr focused. While I'm quite happy to support that workflow as the primary and initial implementation, I want to make sure we leave room for a Rasterio->COG workflow or Pandas->Parquet workflow.